core/modules/file/tests/file.test

```diff
@@ -1773,6 +1773,165 @@ class FilePrivateTestCase extends FileTestHelper {
   }
 }
 
+/**
+ * Confirm that file field submissions work correctly for anonymous visitors.
+ */
+class FileFieldAnonymousSubmission extends FileTestHelper {
+
+  function setUp() {
+    parent::setUp();
+
+    // Allow node submissions by anonymous users.
+    user_role_grant_permissions(BACKDROP_ANONYMOUS_ROLE, array(
+      'create page content',
+      'access content',
+    ));
+  }
+
+  /**
+   * Tests the basic node submission for an anonymous visitor.
+   */
+  function testAnonymousNode() {
+    $bundle_label = 'Page';
+    $node_title = 'Test page';
+
+    // Load the node form.
+    $this->backdropGet('node/add/page');
+    $this->assertResponse(200, 'Loaded the page node form.');
+    $this->assertText(strip_tags(t('Create @name', array('@name' => $bundle_label))));
+
+    $edit = array(
+      'title' => $node_title,
+      'body[und][0][value]' => 'Test page',
+      'body[und][0][format]' => 'filtered_html',
+    );
+    $this->backdropPost(NULL, $edit, t('Save'));
+    $this->assertResponse(200);
+    $t_args = array('@type' => $bundle_label, '%title' => $node_title);
+    $this->assertText(strip_tags(t('@type %title has been created.', $t_args)), 'The node was created.');
+    $matches = array();
+    if (preg_match('@node/(\d+)$@', $this->getUrl(), $matches)) {
+      $nid = end($matches);
+      $this->assertNotEqual($nid, 0, 'The node ID was extracted from the URL.');
+      $node = node_load($nid);
+      $this->assertNotEqual($node, NULL, 'The node was loaded successfully.');
+    }
+  }
+
+  /**
+   * Tests file submission for an anonymous visitor.
+   */
+  function testAnonymousNodeWithFile() {
+    $type_name = 'page';
+    $bundle_label = 'Page';
+    $node_title = 'Test page';
+
+    $field_name = strtolower($this->randomName());
+    $this->createFileField($field_name, $type_name);
+
+    // Load the node form.
+    $this->backdropGet('node/add/' . $type_name);
+    $this->assertResponse(200, 'Loaded the page node form.');
+    $this->assertText(strip_tags(t('Create @name', array('@name' => $bundle_label))));
+
+    // Generate a file.
+    $test_file = $this->getTestFile('text');
+
+    // Submit the form.
+    $edit = array(
+      'title' => $node_title,
+      'body[und][0][value]' => 'Test page',
+      'body[und][0][format]' => 'filtered_html',
+      'files[' . $field_name .'_und_0]' => backdrop_realpath($test_file->uri),
+    );
+    $this->backdropPost(NULL, $edit, t('Save'));
```

Comment:
In order to fully test the scenario laid out by @docwilmot in https://github.com/backdrop/backdrop-issues/issues/1678#issuecomment-885155570, we would need to first submit the form with only the `'files[' . $field_name .'_und_0]'` value, but leave a required field (like title) blank. Confirm the file was uploaded and while a validation error occurred with an assert, then attempt to save the form again after setting a title.

Reply:
@quicksketch the test actually does this already further down in the code.

I've figured out the issue; I needed to make sure the node doesn't have a path alias. Backdrop automatically sets aliases but not Drupal 7. This explains why it wasn't testing it properly in Backdrop when I copied it over from Drupal 7.